cpp/include/cudf/strings/detail/combine.hpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/table/table_view.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * @copydoc concatenate(table_view const&,string_scalar const&,string_scalar
+ * const&,rmm::mr::device_memory_resource*)
+ *
+ * @param stream Stream on which to run kernels
+ */
+std::unique_ptr<column> concatenate(
+  table_view const& strings_columns,
+  string_scalar const& separator      = string_scalar(""),
+  string_scalar const& narep          = string_scalar("", false),
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+  cudaStream_t stream                 = 0);
```

Comment:
After discussing this in https://github.com/rapidsai/cudf/issues/5119, I think we should put `stream` before `mr` and remove the default `0`. Besides the implementation of the public API itself, the detail API only really needs to be used where there's an existing stream, and in other cases the public API can be used instead.

However, removing the default from `stream` would require removing the defaults for `separator` and `narep` from the detail API as well. This is probably not an issue for the same reason as above, as the public API could be used instead in cases where the defaults are desired.

Reply:
I'm not sure what to do here. I'm ok with reordering the stream and mr parameters but not with losing the default parameters in the details function.

Reply:
I changed the parameter order but left the defaults.

Reply:
What's the point of having defaults in the detail API? If you don't have a stream to pass to it, you can just use the non-detail API which has the defaults for `separator` and `narep`. If you _do_ have a stream to pass to it (or a MR for that matter), you wouldn't be able to use those defaults anyway since they come earlier in the parameter list.

Reply:
To put it another way, as I've been browsing through uses of detail APIs while working on #5119, they almost always have at least a `stream` passed in, and in places where they didn't, it was usually a bug where `stream` or `mr` _should_ have been passed in.

If there really should be some defaulted parameters, maybe they could come after `stream` and `mr`, or they could be members of a struct parameter that sets the defaults in its constructor.

Reply:
Ok. I've removed the defaults from these detail functions.